server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java

```diff
@@ -0,0 +1,352 @@
+        );
+
+        ClusterInfo clusterInfo = ClusterInfo.builder().build();
+        routingAllocation = new RoutingAllocation(null, clusterState.getRoutingNodes(), clusterState, clusterInfo, null, System.nanoTime());
+        routingAllocation.setDebugMode(RoutingAllocation.DebugMode.ON);
+
+        indexBalanceAllocationDecider = new IndexBalanceAllocationDecider(builder.build(), createBuiltInClusterSettings(builder.build()));
+
+        indexTierShardRouting = TestShardRouting.newShardRouting(
+            new ShardId(indexMetadata.getIndex(), 1),
+            randomFrom(indexNodeOne, indexNodeTwo).getId(),
+            null,
+            true,
+            ShardRoutingState.STARTED
+        );
+
+        searchTierShardRouting = TestShardRouting.newShardRouting(
+            new ShardId(indexMetadata.getIndex(), 1),
+            randomFrom(searchNodeOne, searchNodeTwo).getId(),
+            null,
+            false,
+            ShardRoutingState.STARTED
+        );
+
+        indexTier = List.of(routingIndexNodeOne, routingIndexNodeTwo);
+        searchIier = List.of(routingSearchNodeOne, routingSearchNodeTwo);
+    }
+
+    public void testCanAllocateUnderThresholdWithExcessShards() {
+        setup(false, true);
+        ShardRouting newIndexShardRouting = TestShardRouting.newShardRouting(
+            new ShardId("newIndex", "uuid", 1),
+            indexNodeTwo.getId(),
+            null,
+            true,
+            ShardRoutingState.STARTED
+        );
+
+        for (RoutingNode routingNode : List.of(
+            routingIndexNodeOne,
+            routingIndexNodeTwo,
+            routingSearchNodeOne,
+            routingSearchNodeTwo,
+            routingMachineLearningNode
+        )) {
+            assertDecisionMatches(
+                "Assigning a new index to a node should succeed",
+                indexBalanceAllocationDecider.canAllocate(newIndexShardRouting, routingNode, routingAllocation),
+                Decision.Type.YES,
+                "Node does not currently host this index."
+            );
+        }
+
+        for (RoutingNode routingNode : searchIier) {
+            assertDecisionMatches(
+                "Assigning a new primary shard to a search tier node should succeed",
+                indexBalanceAllocationDecider.canAllocate(indexTierShardRouting, routingNode, routingAllocation),
+                Decision.Type.YES,
+                "A search node cannot own primary shards. Decider inactive."
+            );
+        }
+
+        for (RoutingNode routingNode : indexTier) {
+            assertDecisionMatches(
+                "Assigning a replica shard to a index tier node should succeed",
+                indexBalanceAllocationDecider.canAllocate(searchTierShardRouting, routingNode, routingAllocation),
+                Decision.Type.YES,
+                "An index node cannot own search shards. Decider inactive."
+            );
+        }
+
+        verifyCanAllocate();
```

Comment:
This seems like the excess shards testing. Unlike the prior checks. Confusing to have the prior checks under testCanAllocateUnderThresholdWithExcessShards.

Maybe make two tests? testCanAllocateUnderThresholdWithExcessShards would have the verifyCanAllocate logic. And a second test, like testAllocateNewIndex.